File: src/main/java/io/github/slimefunguguproject/bump/implementation/items/tools/GetgoldSpade.java
package io.github.slimefunguguproject.bump.implementation.items.tools;

import javax.annotation.Nonnull;

import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

import io.github.slimefunguguproject.bump.implementation.Bump;
import io.github.slimefunguguproject.bump.implementation.BumpItems;
import io.github.slimefunguguproject.bump.implementation.setup.BumpItemGroups;
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType;
import io.github.thebusybiscuit.slimefun4.core.handlers.ToolUseHandler;
import io.github.thebusybiscuit.slimefun4.implementation.SlimefunItems;
import io.github.thebusybiscuit.slimefun4.implementation.items.SimpleSlimefunItem;

import net.guizhanss.guizhanlib.utils.RandomUtil;

/**
 * A {@link GetgoldSpade getting-gold shovel} will give player old coin
 * with chance if player is breaking sand.
 *
 * @author ybw0014
 */
public class GetgoldSpade extends SimpleSlimefunItem<ToolUseHandler> {

    public GetgoldSpade() {
        super(BumpItemGroups.TOOL, BumpItems.GETGOLD_SPADE, RecipeType.ENHANCED_CRAFTING_TABLE, new ItemStack[]{
            null, SlimefunItems.GOLD_24K, null,
            null, new ItemStack(Material.STICK), null,
            null, new ItemStack(Material.STICK), null
        });
    }

    @Nonnull
    @Override
    public ToolUseHandler getItemHandler() {
        return (e, tool, fortune, drops) -> {
            if (e.getBlock().getType() != Material.SAND) {
                return;
            }

            if (!RandomUtil.testChance(1, 4)) {
                return;
            }

            drops.clear();
            drops.add(BumpItems.OLD_COIN);

            Bump.getLocalization().sendActionbarMessage(e.getPlayer(), "tool.getgold_spade");
        };
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/implementation/items/tools/GetGoldSpade.kt
package io.github.slimefunguguproject.bump.implementation.items.tools

import io.github.slimefunguguproject.bump.Bump
import io.github.slimefunguguproject.bump.implementation.BumpItems
import io.github.thebusybiscuit.slimefun4.api.items.ItemGroup
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType
import io.github.thebusybiscuit.slimefun4.core.handlers.ToolUseHandler
import io.github.thebusybiscuit.slimefun4.implementation.items.SimpleSlimefunItem
import net.guizhanss.guizhanlib.utils.RandomUtil
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.inventory.ItemStack
import kotlin.math.min

class GetGoldSpade(
    itemGroup: ItemGroup,
    itemStack: SlimefunItemStack,
    recipeType: RecipeType,
    recipe: Array<ItemStack?>
) : SimpleSlimefunItem<ToolUseHandler>(itemGroup, itemStack, recipeType, recipe) {
    override fun getItemHandler() =
        ToolUseHandler { e: BlockBreakEvent, item: ItemStack, _: Int, drops: MutableList<ItemStack> ->
            if (e.block.type != Material.SAND) return@ToolUseHandler

            val luckLevel = min(item.getEnchantmentLevel(Enchantment.LUCK), 5)
            if (!RandomUtil.testChance(30 + luckLevel * 10, 100)) return@ToolUseHandler

            if (BumpItems.BROKEN_GOLD_COIN.item?.isDisabledIn(e.block.world) == true) return@ToolUseHandler

            drops.clear()
            drops.add(BumpItems.BROKEN_GOLD_COIN)
            Bump.localization.sendActionbarMessage(e.player, "tool.getgold_spade")
        }
}
